pilot/pkg/model/proxy_config.go

```diff
@@ -140,6 +141,9 @@ func mergeWithPrecedence(pcs ...*meshconfig.ProxyConfig) *meshconfig.ProxyConfig
 		if pcs[i].GetImage() != nil {
 			merged.Image = pcs[i].GetImage().DeepCopy()
 		}
+		if pcs[i].GetTracing() != nil {
+			merged.Tracing = proto.Clone(pcs[i].GetTracing()).(*meshconfig.Tracing)
```

Comment:
Ok I spent a long time with this and I think I finally get the issue :slightly_smiling_face: ...

So, mergeWithPrecedence was added when we added the ProxyConfig CRD (not configmap), which has a subset of fields defined. This had merging issues, which we worked around with logic like this. You can see we only have 2 fields, which is because the ProxyConfig CRD only has a few fields anyways so we did them all one-by-one like this.

The problem is mergeWithPrecedence doesn't operate on the Proxyconfig CRD, it operates on the full meshconfig.ProxyConfig, which has 100s of fields. As you found, tracing is one of them that has issues.

We have already established a merging logic for this type: https://github.com/istio/istio/blob/40b352b674327a009776a8a66c12bc51d9f26145/pkg/config/mesh/mesh.go#L144 which correctly handles the concurrency, image, tracing and every other case.

IMO the proper fix would be to use that logic. I did a small protoype which seems to work, but needs some cleaning up etc:


```diff
diff --git a/pilot/pkg/model/proxy_config.go b/pilot/pkg/model/proxy_config.go
index f54f477350..7956ee1e3c 100644
--- a/pilot/pkg/model/proxy_config.go
+++ b/pilot/pkg/model/proxy_config.go
@@ -15,15 +15,12 @@
 package model
 
 import (
-	wrappers "google.golang.org/protobuf/types/known/wrapperspb"
-
 	"istio.io/api/annotation"
 	meshconfig "istio.io/api/mesh/v1alpha1"
 	"istio.io/api/networking/v1beta1"
 	"istio.io/istio/pkg/config/labels"
 	"istio.io/istio/pkg/config/mesh"
 	"istio.io/istio/pkg/config/schema/collections"
-	"istio.io/istio/pkg/proto/merge"
 	"istio.io/istio/pkg/util/protomarshal"
 	istiolog "istio.io/pkg/log"
 )
@@ -133,13 +130,15 @@ func mergeWithPrecedence(pcs ...*meshconfig.ProxyConfig) *meshconfig.ProxyConfig
 		// TODO(Monkeyanator) some fields seem not to merge when set to the type's default value
 		// such as overriding with a concurrency value 0. Do we need a custom merge similar to what the
 		// telemetry code does with shallowMerge?
-		merge.Merge(merged, pcs[i])
-		if pcs[i].GetConcurrency() != nil {
-			merged.Concurrency = wrappers.Int32(pcs[i].GetConcurrency().GetValue())
-		}
-		if pcs[i].GetImage() != nil {
-			merged.Image = pcs[i].GetImage().DeepCopy()
-		}
+		n, _ := protomarshal.ToYAML(pcs[i])
+		merged,_ = mesh.ApplyProxyConfig2(n, merged)
+		//merge.Merge(merged, pcs[i])
+		//if pcs[i].GetConcurrency() != nil {
+		//	merged.Concurrency = wrappers.Int32(pcs[i].GetConcurrency().GetValue())
+		//}
+		//if pcs[i].GetImage() != nil {
+		//	merged.Image = pcs[i].GetImage().DeepCopy()
+		//}
 	}
 	return merged
 }
diff --git a/pkg/config/mesh/mesh.go b/pkg/config/mesh/mesh.go
index 788ed29e23..797e49813d 100644
--- a/pkg/config/mesh/mesh.go
+++ b/pkg/config/mesh/mesh.go
@@ -148,6 +148,14 @@ func ApplyProxyConfig(yaml string, meshConfig *meshconfig.MeshConfig) (*meshconf
 	mc.DefaultConfig = pc
 	return mc, nil
 }
+func ApplyProxyConfig2(yaml string, defaultConfig *meshconfig.ProxyConfig) (*meshconfig.ProxyConfig, error) {
+	mc := proto.Clone(defaultConfig).(*meshconfig.ProxyConfig)
+	pc, err := applyProxyConfig(yaml, mc)
+	if err != nil {
+		return nil, err
+	}
+	return pc, nil
+}
 
 func applyProxyConfig(yaml string, proxyConfig *meshconfig.ProxyConfig) (*meshconfig.ProxyConfig, error) {
 	origMetadata := proxyConfig.ProxyMetadata
```

Reply:
Thank you for digging deeper! Your suggested fix sounds reasonable. I'll check and come back soon 😉